### PR TITLE
[ UNITTEST ] Integer Unit test for conv2d

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR inlucdes Integet Unit Test for conv2d to evaluted accuracy.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: test/input_gen/genInput.py [hunk 1]
```diff
@@ -51,6 +51,12 @@ def save(filename, *data):
 def gen_tensor(shape, dtype=dtypes.float32):
     return np.random.randint(1, 10, shape).astype(np.float32)
 
+##
+# @brief generate random tensor
+def gen_tensor_int(shape):
+  rand_t = np.random.randint(1, 100, size=shape)
```
Comment (kparichay) [inline test/input_gen/genInput.py:57]: please limit max value to 10. 
Values to 100 causes overflow with bigger layers/values.

Comment (kparichay) [reply test/input_gen/genInput.py:57]: Also, similar values can bring out errors like from pooling etc.

File: test/input_gen/genInput.py [hunk 2]
```diff
@@ -93,22 +102,73 @@ def conv2d_tf(x, kernel, batch, width, height, channel, k_width, k_height, k_num
...
-    grad = tf.gradients(conv2d_layer, input_variables)
+    grad = tf.gradients(conv2d_layer, all_variables)
+    train_op = optimizer.apply_gradients(list(zip(grad[1:], trainable_variables)))
 
     with tf.compat.v1.Session() as sess:
         sess.run(tf.global_variables_initializer())
-        conv2d_result = sess.run(conv2d_layer, feed_dict={tf_input: x})
-        grad_result = sess.run(grad, feed_dict={tf_input:x})
+
+        conv2d_result, grad_result, _ = sess.run([conv2d_layer, grad, train_op], feed_dict={tf_input: x})
```
Comment (kparichay) [inline test/input_gen/genInput.py:116]: take an input for number of iterations. Then run this in a loop.
This allows the earlier unittest - `conv2d_1` and `conv2d_2` which also use this function to be reproducible

File: test/input_gen/genInput.py [hunk 3]
```diff
@@ -51,6 +51,12 @@ def save(filename, *data):
 def gen_tensor(shape, dtype=dtypes.float32):
     return np.random.randint(1, 10, shape).astype(np.float32)
 
+##
+# @brief generate random tensor
+def gen_tensor_int(shape):
```
Comment (kparichay) [inline test/input_gen/genInput.py:56]: How is this different from `gen_tensor`? Is this just for max value to be 100?

Comment (kparichay) [reply test/input_gen/genInput.py:56]: I think you will have to remove int from the gen_tensor  and limit range here to be 10.

File: test/input_gen/genInput.py [hunk 4]
```diff
@@ -351,16 +411,55 @@ def bn_tf(x, *, trainable=True, init_beta=gen_tensor, init_gamma=gen_tensor, axi
...
     save(base_name+"_goldenConv2DResult.out", np.transpose(golden_conv,(0,3,1,2)))
     save(base_name+"_goldenInputGrad.out", np.transpose(golden_grad_input,(0,3,1,2)))
     save(base_name+"_goldenKernelGrad.out", np.transpose(golden_grad_kernel,(3,2,0,1)))
     save(base_name+"_goldenBiasGrad.out", golden_grad_bias)
+    save(base_name+"_goldenConv2DResult2.out", np.transpose(golden_conv2,(0,3,1,2)))
+    save(base_name+"_goldenInputGrad2.out", np.transpose(golden_grad_input2,(0,3,1,2)))
+    save(base_name+"_goldenKernelGrad2.out", np.transpose(golden_grad_kernel2,(3,2,0,1)))
+    save(base_name+"_goldenBiasGrad2.out", golden_grad_bias2)
+
+def gen_test_case_conv_int(base_name):
```
Comment (kparichay) [inline test/input_gen/genInput.py:431]: as all the unittests now use int like values, rename this to `gen_test_case_conv_multi_iter`

File: test/input_gen/genInput.py [hunk 5]
```diff
@@ -493,6 +592,12 @@ def gen_test_case_bn(input_shape, base_name, training=True):
         bias2 = [1.0, 1.0, 1.0]
         gen_test_case_conv(2, 3, 7, 7, 3, 3, 3, "VALID", 1, bias2, "tc_conv2d_2")
 
+    if target == "conv2d_int":
```
Comment (kparichay) [inline test/input_gen/genInput.py:595]: same as above
rename to conv2d_2layers

File: test/input_gen/genInput.py [hunk 6]
```diff
@@ -351,16 +411,55 @@ def bn_tf(x, *, trainable=True, init_beta=gen_tensor, init_gamma=gen_tensor, axi
...
+    save(base_name+"_goldenBias2Grad.out", golden_grads[4])
+
+def gen_test_case_conv_3(base_name):
+    x=gen_input(base_name+"_conv2DLayer.in", [1, 6, 24, 24], int_data=True)
+    kernel=gen_input(base_name+"_conv2DKernel.in", [12, 6, 5, 5], int_data=True)
+    bias = np.ones(12)
+    with open(base_name+"_conv2DKernel.in", 'ab') as outfile:
+        np.array(bias, dtype=np.float32).tofile(outfile)
+
+    golden_conv, golden_grads = conv2d_tf_2(x, kernel, bias, kernel, bias)
```
Comment (kparichay) [inline test/input_gen/genInput.py:458]: I think this isnt correct
This function returns 3 values

Comment (kparichay) [inline test/input_gen/genInput.py:458]: I think this unittest can just be removed
it doesn't do anything new

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 7]
```diff
@@ -1027,6 +1028,59 @@ TEST_F(nntrainer_Conv2DLayer, backwarding_01_p) {
   matchOutput(bias_grad, "tc_conv2d_1_goldenBiasGrad.out");
 }
 
+TEST_F(nntrainer_Conv2DLayer, backwarding_0r_p) {
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:1031]: backwarding_04_p?

**Review by zhoonit — APPROVED** [review]
LGTM after above comment has been resolved.

File: test/input_gen/genInput.py (hunk 4 above)
Comment (jijoongmoon) [reply test/input_gen/genInput.py]: fixed. thanks.

File: test/input_gen/genInput.py (hunk 2 above)
Comment (jijoongmoon) [reply test/input_gen/genInput.py:106]: fixed. thanks.

File: test/input_gen/genInput.py (hunk 6 above)
Comment (jijoongmoon) [reply test/input_gen/genInput.py:446]: right.

File: test/input_gen/genInput.py (hunk 5 above)
Comment (jijoongmoon) [reply test/input_gen/genInput.py]: done.

File: test/input_gen/genInput.py [hunk 8]
```diff
@@ -93,22 +92,71 @@ def conv2d_tf(x, kernel, batch, width, height, channel, k_width, k_height, k_num
...
+    grad = tf.gradients(conv2d_layer, all_variables)
+    train_op = optimizer.apply_gradients(list(zip(grad[1:], trainable_variables)))
 
     with tf.compat.v1.Session() as sess:
         sess.run(tf.global_variables_initializer())
-        conv2d_result = sess.run(conv2d_layer, feed_dict={tf_input: x})
-        grad_result = sess.run(grad, feed_dict={tf_input:x})
+
+        conv2d_result, grad_result, _ = sess.run([conv2d_layer, grad, train_op], feed_dict={tf_input: x})
+        for i in range(0:4):
```
Comment (kparichay) [inline test/input_gen/genInput.py:107]: the max loop range needs to takes as an argument with default 1 to support old unittest cases reproduction.

Comment (jijoongmoon) [reply test/input_gen/genInput.py]: fixed. thanks.

**Review by kparichay — APPROVED** [review]
LGTM

**Pull request merged**